velox/functions/sparksql/specialforms/FromJson.cpp

```diff
@@ -99,7 +101,63 @@ struct ExtractJsonTypeImpl {
   struct KindDispatcher<TypeKind::INTEGER, Dummy> {
     static simdjson::error_code
     apply(Input value, exec::GenericWriter& writer, bool /*isRoot*/) {
-      return castJsonToInt<int32_t>(value, writer);
+      if (writer.type() == DATE()) {
+        SIMDJSON_ASSIGN_OR_RAISE(auto type, value.type());
+        std::string_view s;
+        switch (type) {
+          case simdjson::ondemand::json_type::string: {
+            SIMDJSON_ASSIGN_OR_RAISE(s, value.get_string());
+            break;
+          }
+          default:
+            return simdjson::INCORRECT_TYPE;
+        }
+        int32_t day = 0;
+        // If there are fewer than four digits, the value is treated as the
+        // number of days since 1970-01-01.
+        if (s.size() < 4) {
+          auto result =
+              folly::tryTo<int32_t>(std::string_view(s.data(), s.size()));
+          if (!result.hasError()) {
+            day = result.value();
+          } else {
+            return simdjson::INCORRECT_TYPE;
+          }
+        } else {
+          auto days =
+              util::fromDateString(StringView(s), util::ParseMode::kSparkCast);
+          if (!days.hasError()) {
+            day = days.value();
+          } else {
+            if (stringImpl::stringPosition<true /*isAscii*/>(
```

Comment:
Do we need to find the GMT and then replace, can we replace directly?

Reply:
The following replace function does not perform in-place replacement. A copy is always made, which is why it first checks whether replacement is needed.

I have tried in-place replacement, but it seems that data copying is unavoidable.
https://github.com/facebookincubator/velox/blob/9f993aa6c927620cac01bcd0b96f92edcd81ab3e/velox/functions/lib/string/tests/StringImplTest.cpp#L518